File: src/codes/Users_Manual/link/soem_0.rs
use autd3::prelude::*;
use autd3_link_soem::{SOEM, TimerStrategy, Status};

# #[allow(unused_variables)]
# #[tokio::main]
# async fn main() -> Result<(), Box<dyn std::error::Error>> {
# let autd = Controller::builder([AUTD3::new(Vector3::zeros())])
#            .open(
SOEM::builder()
    .with_ifname("")
    .with_buf_size(32)
    .with_err_handler(|slave, status| match status {
        Status::Error => eprintln!("Error [{}]: {}", slave, status),
        Status::Lost => {
            eprintln!("Lost [{}]: {}", slave, status);
            // You can also wait for the link to recover, without exitting the process
            std::process::exit(-1);
        }
        Status::StateChanged => eprintln!("StateChanged [{}]: {}", slave, status),
    })
    .with_state_check_interval(std::time::Duration::from_millis(100))
    .with_sync0_cycle(2)
    .with_send_cycle(2)
    .with_timer_strategy(TimerStrategy::BusyWait)
    .with_sync_tolerance(std::time::Duration::from_micros(1))
    .with_sync_timeout(std::time::Duration::from_secs(10))
# ).await?;
# Ok(())
# }

File: src/codes/Users_Manual/link/soem_0.py
~import os
~from datetime import timedelta
from pyautd3.link.soem import SOEM, Status, TimerStrategy


def err_handler(slave: int, status: Status, msg: str) -> None:
    match status:
        case Status.Error:
            print(f"Error [{slave}]: {msg}")
        case Status.Lost:
            print(f"Lost [{slave}]: {msg}")
            # You can also wait for the link to recover, without exiting the process
            os._exit(-1)
        case Status.StateChanged:
            print(f"StateChanged  [{slave}]: {msg}")


SOEM.builder()\
    .with_ifname("")\
    .with_buf_size(32)\
    .with_err_handler(
        err_handler
    )\
    .with_state_check_interval(timedelta(milliseconds=100))\
    .with_sync0_cycle(2)\
    .with_send_cycle(2)\
    .with_timer_strategy(TimerStrategy.BusyWait)\
    .with_sync_tolerance(timedelta(microseconds=1))\
    .with_sync_timeout(timedelta(seconds=10))